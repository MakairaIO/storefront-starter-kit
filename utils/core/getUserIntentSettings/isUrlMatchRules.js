function operatorComparision(operator, sourceValue, compareValue = '') {
  try {
    switch (operator) {
      case 'is':
        if (sourceValue === compareValue) return true
        break
      case 'contains':
        if (sourceValue.includes(compareValue)) return true
        break
      case 'begins':
        if (sourceValue.startsWith(compareValue)) return true
        break
      case 'wildcard':
        {
          const regex = new RegExp(`^${compareValue.replace('*', '.*')}`, 'g')
          if (sourceValue.match(regex)) return true
        }
        break
      default:
        break
    }
  } catch (error) {
    console.debug(error)
  }
  return false
}

function stripTrailingSlash(str) {
  return str.replace(/\/$/, '')
}

export function isUrlMatchRules(rules) {
  const url = stripTrailingSlash(window.location.pathname)
  const searchStr = window.location.search.replace('?', '')
  const query = Object.fromEntries(new URLSearchParams(searchStr).entries())
  for (let rule of rules) {
    let isMatched = false
    if (rule.key === 'url') {
      isMatched = operatorComparision(
        rule.operator,
        url,
        stripTrailingSlash(rule.value)
      )
    }
    if (rule.key === 'query') {
      const queryValue = rule.queryName && query[rule.queryName]
      isMatched =
        queryValue !== undefined
          ? operatorComparision(rule.operator, queryValue, rule.value)
          : false
    }

    if (isMatched) return true
  }
  return false
}
