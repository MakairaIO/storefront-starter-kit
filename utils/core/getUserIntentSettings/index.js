import { isUrlMatchRules } from './isUrlMatchRules'

const CTA_TYPE = Object.freeze({
  SLIDEIN: 'slidein',
  POPUP: 'popup',
})

const reduceLatest = (prev, curr) => {
  if (!prev) return curr
  const a = new Date(prev.updatedAt)
  const b = new Date(curr.updatedAt)
  if (a.getTime() - b.getTime() >= 0) return prev
  else return curr
}

function filterLatestUpdateDocument(scenarioDocuments = [], documents = []) {
  const mapping = documents
    .map((id) => {
      const document = scenarioDocuments.find((item) => item.id === id)
      if (document) {
        const [ctaType] = document.ctaType.split('_')
        return {
          id,
          updatedAt: document.updatedAt,
          ctaType,
        }
      }
      return null
    })
    .filter((item) => !!item)

  const slideIn = mapping
    .filter((item) => item.ctaType === CTA_TYPE.SLIDEIN)
    .reduce(reduceLatest, null)
  const popup = mapping
    .filter((item) => item.ctaType === CTA_TYPE.POPUP)
    .reduce(reduceLatest, null)
  const result = []
  if (slideIn) result.push(slideIn)
  if (popup) result.push(popup)
  return result
}
/**
 * Return {
    pageScroll: [
      { value: 40, documents: [] },
      { value: 60, documents: [''] }
    ],
    pageExit: [ '', '' ],
    pageInactivity: [
      { value: 5, documents: [''] },
      { value: 5, documents: [''] }
    ],
    pageElapsed: [
      { value: 5, documents: [''] },
      { value: 10, documents: [''] }
    ]
  }
  */
export default function getUserIntentSettings(documents = []) {
  const initData = {
    pageScroll: [],
    pageExit: [],
    pageInactivity: [],
    pageElapsed: [],
  }

  for (let document of documents) {
    const { setting = {}, id, active } = document
    const exclude = document?.urlRule?.exclude || []
    const or = document?.urlRule?.or || []

    if (!active) continue

    const isMatchedExcludeRule = isUrlMatchRules(exclude)

    if (!isMatchedExcludeRule) {
      if (isUrlMatchRules(or)) {
        for (let settingKey in setting) {
          const updating = initData[settingKey]
          switch (settingKey) {
            case 'pageExit':
              if (setting[settingKey]) updating.push(id)
              break
            case 'pageScroll':
            case 'pageElapsed':
            case 'pageInactivity':
              {
                const index = updating.findIndex(
                  (item) => item.value === setting[settingKey]
                )
                if (index > -1) {
                  updating[index].documents.push(id)
                } else {
                  updating.push({
                    value: setting[settingKey],
                    documents: [id],
                  })
                }
              }
              break
            default:
              break
          }
        }
      }
    }
  }

  for (let key of Object.keys(initData)) {
    switch (key) {
      case 'pageExit':
        initData[key] = filterLatestUpdateDocument(documents, initData[key])
        break
      case 'pageScroll':
      case 'pageElapsed':
      case 'pageInactivity':
        {
          for (let item of initData[key]) {
            item.documents = filterLatestUpdateDocument(
              documents,
              item.documents
            )
          }
        }
        break
      default:
        break
    }
  }

  return initData
}
