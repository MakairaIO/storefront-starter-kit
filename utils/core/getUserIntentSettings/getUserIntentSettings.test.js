import getUserIntentSettings from './'
import {
  MultiUserIntentWithDifferentSettings,
  MultiUserIntentWithSameSettingsDifferentCTAType,
  MultiUserIntentWithSameSettingsSameCTAType,
  SingleUserIntentWithAllSettings,
  SingleUserIntentWithFewSettings,
} from './mocks/userIntentSettings'

describe('getUserIntentSettings()', () => {
  beforeEach(() => {
    const url = 'http://dummy.com'
    Object.defineProperty(window, 'location', {
      value: {
        href: url,
        pathname: '/',
        search: '',
      },
      writable: true, // possibility to override
    })
  })

  it('Should return right structure single intent', () => {
    const result = getUserIntentSettings(SingleUserIntentWithAllSettings)
    expect(result).toMatchSnapshot()
  })

  it('Should return right structure single intent 2', () => {
    const result = getUserIntentSettings(SingleUserIntentWithFewSettings)
    expect(result).toMatchSnapshot()
  })

  it('Should return right structure multi intent different setting', () => {
    const result = getUserIntentSettings(MultiUserIntentWithDifferentSettings)
    expect(result).toMatchSnapshot()
  })

  it('Should return right structure multi intent same setting same CTAType', () => {
    const result = getUserIntentSettings(
      MultiUserIntentWithSameSettingsSameCTAType
    )
    expect(result).toMatchSnapshot()
  })

  it('Should return right structure multi intent same setting different CTAType', () => {
    const result = getUserIntentSettings(
      MultiUserIntentWithSameSettingsDifferentCTAType
    )
    expect(result).toMatchSnapshot()
  })
})
