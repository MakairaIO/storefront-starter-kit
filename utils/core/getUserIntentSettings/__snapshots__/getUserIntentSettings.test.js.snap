// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserIntentSettings() Should return right structure multi intent different setting 1`] = `
Object {
  "pageElapsed": Array [
    Object {
      "documents": Array [
        Object {
          "ctaType": "slidein",
          "id": "2",
          "updatedAt": "2023-07-31T13:41:07+00:00",
        },
      ],
      "value": 10,
    },
  ],
  "pageExit": Array [
    Object {
      "ctaType": "slidein",
      "id": "1",
      "updatedAt": "2023-08-03T14:52:20+00:00",
    },
  ],
  "pageInactivity": Array [
    Object {
      "documents": Array [
        Object {
          "ctaType": "slidein",
          "id": "2",
          "updatedAt": "2023-07-31T13:41:07+00:00",
        },
      ],
      "value": 20,
    },
  ],
  "pageScroll": Array [
    Object {
      "documents": Array [
        Object {
          "ctaType": "slidein",
          "id": "1",
          "updatedAt": "2023-08-03T14:52:20+00:00",
        },
      ],
      "value": 40,
    },
  ],
}
`;

exports[`getUserIntentSettings() Should return right structure multi intent same setting different CTAType 1`] = `
Object {
  "pageElapsed": Array [],
  "pageExit": Array [
    Object {
      "ctaType": "slidein",
      "id": "2",
      "updatedAt": "2023-07-31T13:41:07+00:00",
    },
    Object {
      "ctaType": "popup",
      "id": "1",
      "updatedAt": "2023-08-03T14:52:20+00:00",
    },
  ],
  "pageInactivity": Array [],
  "pageScroll": Array [
    Object {
      "documents": Array [
        Object {
          "ctaType": "slidein",
          "id": "2",
          "updatedAt": "2023-07-31T13:41:07+00:00",
        },
        Object {
          "ctaType": "popup",
          "id": "1",
          "updatedAt": "2023-08-03T14:52:20+00:00",
        },
      ],
      "value": 40,
    },
  ],
}
`;

exports[`getUserIntentSettings() Should return right structure multi intent same setting same CTAType 1`] = `
Object {
  "pageElapsed": Array [],
  "pageExit": Array [
    Object {
      "ctaType": "slidein",
      "id": "1",
      "updatedAt": "2023-08-03T14:52:20+00:00",
    },
  ],
  "pageInactivity": Array [],
  "pageScroll": Array [
    Object {
      "documents": Array [
        Object {
          "ctaType": "slidein",
          "id": "1",
          "updatedAt": "2023-08-03T14:52:20+00:00",
        },
      ],
      "value": 40,
    },
  ],
}
`;

exports[`getUserIntentSettings() Should return right structure single intent 1`] = `
Object {
  "pageElapsed": Array [
    Object {
      "documents": Array [
        Object {
          "ctaType": "slidein",
          "id": "1",
          "updatedAt": "2023-08-03T14:52:20+00:00",
        },
      ],
      "value": 10,
    },
  ],
  "pageExit": Array [
    Object {
      "ctaType": "slidein",
      "id": "1",
      "updatedAt": "2023-08-03T14:52:20+00:00",
    },
  ],
  "pageInactivity": Array [
    Object {
      "documents": Array [
        Object {
          "ctaType": "slidein",
          "id": "1",
          "updatedAt": "2023-08-03T14:52:20+00:00",
        },
      ],
      "value": 20,
    },
  ],
  "pageScroll": Array [
    Object {
      "documents": Array [
        Object {
          "ctaType": "slidein",
          "id": "1",
          "updatedAt": "2023-08-03T14:52:20+00:00",
        },
      ],
      "value": 40,
    },
  ],
}
`;

exports[`getUserIntentSettings() Should return right structure single intent 2 1`] = `
Object {
  "pageElapsed": Array [],
  "pageExit": Array [
    Object {
      "ctaType": "slidein",
      "id": "1",
      "updatedAt": "2023-08-03T14:52:20+00:00",
    },
  ],
  "pageInactivity": Array [],
  "pageScroll": Array [
    Object {
      "documents": Array [
        Object {
          "ctaType": "slidein",
          "id": "1",
          "updatedAt": "2023-08-03T14:52:20+00:00",
        },
      ],
      "value": 40,
    },
  ],
}
`;
